File: application/src/main/java/org/thingsboard/server/controller/CustomTranslationController.java
/**
 * ThingsBoard, Inc. ("COMPANY") CONFIDENTIAL
 *
 * Copyright © 2016-2024 ThingsBoard, Inc. All Rights Reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of ThingsBoard, Inc. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to ThingsBoard, Inc.
 * and its suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 *
 * Dissemination of this information or reproduction of this material is strictly forbidden
 * unless prior written permission is obtained from COMPANY.
 *
 * Access to the source code contained herein is hereby forbidden to anyone except current COMPANY employees,
 * managers or contractors who have executed Confidentiality and Non-disclosure agreements
 * explicitly covering such access.
 *
 * The copyright notice above does not evidence any actual or intended publication
 * or disclosure  of  this source code, which includes
 * information that is confidential and/or proprietary, and is a trade secret, of  COMPANY.
 * ANY REPRODUCTION, MODIFICATION, DISTRIBUTION, PUBLIC  PERFORMANCE,
 * OR PUBLIC DISPLAY OF OR THROUGH USE  OF THIS  SOURCE CODE  WITHOUT
 * THE EXPRESS WRITTEN CONSENT OF COMPANY IS STRICTLY PROHIBITED,
 * AND IN VIOLATION OF APPLICABLE LAWS AND INTERNATIONAL TREATIES.
 * THE RECEIPT OR POSSESSION OF THIS SOURCE CODE AND/OR RELATED INFORMATION
 * DOES NOT CONVEY OR IMPLY ANY RIGHTS TO REPRODUCE, DISCLOSE OR DISTRIBUTE ITS CONTENTS,
 * OR TO MANUFACTURE, USE, OR SELL ANYTHING THAT IT  MAY DESCRIBE, IN WHOLE OR IN PART.
 */
package org.thingsboard.server.controller;

import com.fasterxml.jackson.databind.JsonNode;
import io.swagger.v3.oas.annotations.Parameter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestPart;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;
import org.thingsboard.common.util.JacksonUtil;
import org.thingsboard.server.common.data.exception.ThingsboardException;
import org.thingsboard.server.common.data.permission.Operation;
import org.thingsboard.server.common.data.permission.Resource;
import org.thingsboard.server.common.data.security.Authority;
import org.thingsboard.server.common.data.translation.CustomTranslation;
import org.thingsboard.server.config.annotations.ApiOperation;
import org.thingsboard.server.dao.service.DataValidator;
import org.thingsboard.server.dao.translation.CustomTranslationService;
import org.thingsboard.server.queue.util.TbCoreComponent;
import org.thingsboard.server.service.security.model.SecurityUser;
import org.thingsboard.server.service.translation.TbCustomTranslationService;

import static org.thingsboard.server.controller.ControllerConstants.MARKDOWN_CODE_BLOCK_END;
import static org.thingsboard.server.controller.ControllerConstants.MARKDOWN_CODE_BLOCK_START;
import static org.thingsboard.server.service.translation.DefaultTranslationService.DEFAULT_LOCALE_KEYS;

@RestController
@TbCoreComponent
@RequestMapping("/api")
public class CustomTranslationController extends BaseController {

    private static final String CUSTOM_TRANSLATION_EXAMPLE = "\n\n" +
            MARKDOWN_CODE_BLOCK_START +
            "{\"home\":\"MyHome\"}" +
            MARKDOWN_CODE_BLOCK_END;

    private static final String CUSTOM_TRANSLATION_PATCH_EXAMPLE = "\n\n" +
            MARKDOWN_CODE_BLOCK_START +
            "{\"notification.active\":\"active\"}" +
            MARKDOWN_CODE_BLOCK_END;

    @Autowired
    private TbCustomTranslationService tbCustomTranslationService;

    @Autowired
    private CustomTranslationService customTranslationService;

    @ApiOperation(value = "Get end-user Custom Translation configuration (getMergedCustomTranslation)",
            notes = "Fetch end-user Custom Translation for specified locale. The custom translation is configured in the white labeling parameters. " +
                    "If custom translation translation is defined on the tenant level, it overrides the custom translation of the system level. " +
                    "Similar, if the custom translation is defined on the customer level, it overrides the translation configuration of the tenant level."
            )
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @GetMapping(value = "/translation/custom/merged/{localeCode}")
    @ResponseBody
    public JsonNode getMergedCustomTranslation(@Parameter(description = "Locale code (e.g. 'en_US').")
                                               @PathVariable("localeCode") String localeCode) throws ThingsboardException {
        Authority authority = getCurrentUser().getAuthority();
        if (Authority.SYS_ADMIN.equals(authority)) {
            return customTranslationService.getCurrentCustomTranslation(getCurrentUser().getTenantId(), getCurrentUser().getCustomerId(), localeCode);
        } else if (Authority.TENANT_ADMIN.equals(authority)) {
            return customTranslationService.getMergedTenantCustomTranslation(getCurrentUser().getTenantId(), localeCode);
        } else if (Authority.CUSTOMER_USER.equals(authority)) {
            return customTranslationService.getMergedCustomerCustomTranslation(getCurrentUser().getTenantId(), getCurrentUser().getCustomerId(), localeCode);
        }
        return JacksonUtil.newObjectNode();
    }

    @ApiOperation(value = "Get Custom Translation configuration (getCustomTranslation)",
            notes = "Fetch the Custom Translation for specified locale that corresponds to the authority of the user. " +
                    "The API call is designed to load the custom translation items for edition. " +
                    "So, the result is NOT merged with the parent level configuration. " +
                    "Let's assume there is a custom translation configured on a system level. " +
                    "And there is no custom translation items configured on a tenant level. " +
                    "In such a case, the API call will return empty object for the tenant administrator. " +
                    "\n\n Response example: " + CUSTOM_TRANSLATION_EXAMPLE +
                    ControllerConstants.WL_READ_CHECK
            )
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @GetMapping(value = "/translation/custom/{localeCode}")
    public JsonNode getCustomTranslation(@Parameter(description = "Locale code (e.g. 'en_US').")
                                                  @PathVariable("localeCode") String localeCode) throws ThingsboardException {
        checkWhiteLabelingPermissions(Operation.READ);
        SecurityUser currentUser = getCurrentUser();
        return customTranslationService.getCurrentCustomTranslation(currentUser.getTenantId(), getCurrentUser().getCustomerId(), localeCode);
    }

    @ApiOperation(value = "Create Or Update Custom Translation (saveCustomTranslation)",
            notes = "Creates or Updates the Custom Translation for specified locale." +
                    "\n\n Request example: " + CUSTOM_TRANSLATION_EXAMPLE +
                    ControllerConstants.WL_WRITE_CHECK)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @PostMapping(value = "/translation/custom/{localeCode}")
    @ResponseStatus(value = HttpStatus.OK)
    public void saveCustomTranslation(@Parameter(description = "Locale code (e.g. 'en_US').")
                                                   @PathVariable("localeCode") String localeCode,
                                                   @Parameter(description = "A JSON value representing the custom translation. See API call notes above for valid example.")
                                                   @RequestBody JsonNode customTranslationValue) throws ThingsboardException {
        checkWhiteLabelingPermissions(Operation.WRITE);
        DataValidator.validateLocaleCode(localeCode);
        DataValidator.validateCustomTranslationKeys(DEFAULT_LOCALE_KEYS, customTranslationValue);
        CustomTranslation customTranslation = CustomTranslation.builder()
                .tenantId(getCurrentUser().getTenantId())
                .customerId(getCurrentUser().getCustomerId())
                .localeCode(localeCode)
                .value(customTranslationValue)
                .build();
        tbCustomTranslationService.saveCustomTranslation(customTranslation);
    }

    @ApiOperation(value = "Update Custom Translation for specified translation keys only (patchCustomTranslation)",
            notes = "The API call is designed to update the custom translation for specified key only. " +
                    "\n\n Request example: " + CUSTOM_TRANSLATION_PATCH_EXAMPLE +
                    ControllerConstants.WL_WRITE_CHECK)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @PatchMapping(value = "/translation/custom/{localeCode}")
    @ResponseStatus(value = HttpStatus.OK)
    public void patchCustomTranslation(@Parameter(description = "Locale code (e.g. 'en_US').")
                                                    @PathVariable("localeCode") String localeCode,
                                                    @Parameter(description = "A JSON value representing the custom translation. See API call notes above for valid example.")
                                                    @RequestBody JsonNode newCustomTranslation) throws ThingsboardException {
        checkWhiteLabelingPermissions(Operation.WRITE);
        DataValidator.validateLocaleCode(localeCode);
        DataValidator.validateCustomTranslationPatch(newCustomTranslation);
        SecurityUser currentUser = getCurrentUser();
        tbCustomTranslationService.patchCustomTranslation(currentUser.getTenantId(), currentUser.getCustomerId(),
                localeCode, newCustomTranslation);
    }

    @ApiOperation(value = "Delete specified key of Custom Translation (deleteCustomTranslationKey) ",
            notes = "The API call is designed to delete specified key of the custom translation and return as a result parent translation." +
                    "(e.g. if tenant translation for key is 'value1' and customer translation is 'value2' then by deletinf key onn customer level you will get 'value1' in response) " +
                    ControllerConstants.WL_WRITE_CHECK)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @DeleteMapping(value = "/translation/custom/{localeCode}/{keyPath}")
    @ResponseStatus(value = HttpStatus.OK)
    public void deleteCustomTranslationKey(@Parameter(description = "Locale code (e.g. 'en_US').")
                                                        @PathVariable("localeCode") String localeCode,
                                                        @Parameter(description = "A string value representing key of the custom translation (e.g. 'notification.active').")
                                                        @PathVariable String keyPath) throws ThingsboardException {
        checkWhiteLabelingPermissions(Operation.WRITE);
        DataValidator.validateLocaleCode(localeCode);
        SecurityUser currentUser = getCurrentUser();
        tbCustomTranslationService.deleteCustomTranslationKey(currentUser.getTenantId(), currentUser.getCustomerId(),
                localeCode, keyPath);
    }

    @ApiOperation(value = "Upload Custom Translation (uploadCustomTranslation)",
            notes = "Upload the Custom Translation for specified locale." +
                    "\n\n Request example: " + CUSTOM_TRANSLATION_EXAMPLE +
                    ControllerConstants.WL_WRITE_CHECK)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @PostMapping(value = "/translation/custom/{localeCode}/upload", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    @ResponseStatus(value = HttpStatus.OK)
    public void uploadCustomTranslation(@Parameter(description = "Locale code (e.g. 'en_US').")
                                                     @PathVariable("localeCode") String localeCode,
                                                     @RequestPart MultipartFile file) throws Exception {
        checkWhiteLabelingPermissions(Operation.WRITE);
        DataValidator.validateLocaleCode(localeCode);

        SecurityUser user = getCurrentUser();
        CustomTranslation customTranslation = new CustomTranslation();
        customTranslation.setTenantId(user.getTenantId());
        customTranslation.setCustomerId(user.getCustomerId());
        customTranslation.setLocaleCode(localeCode);
        customTranslation.setValueBytes(file.getBytes());

        tbCustomTranslationService.saveCustomTranslation(customTranslation);
    }

    @ApiOperation(value = "Delete Custom Translation for specified locale (deleteCustomTranslation)",
            notes = "Delete entire custom translation settings for end-user" +
                    ControllerConstants.WL_WRITE_CHECK)
    @PreAuthorize("hasAnyAuthority('SYS_ADMIN', 'TENANT_ADMIN', 'CUSTOMER_USER')")
    @DeleteMapping(value = "/translation/custom/{localeCode}")
    @ResponseStatus(value = HttpStatus.OK)
    public void deleteCustomTranslation(@Parameter(description = "Locale code (e.g. 'en_US').")
                                        @PathVariable("localeCode") String localeCode) throws ThingsboardException {
        checkWhiteLabelingPermissions(Operation.WRITE);
        DataValidator.validateLocaleCode(localeCode);
        SecurityUser currentUser = getCurrentUser();
        tbCustomTranslationService.deleteCustomTranslation(currentUser.getTenantId(), currentUser.getCustomerId(),
                localeCode);
    }

    private void checkWhiteLabelingPermissions(Operation operation) throws ThingsboardException {
        accessControlService.checkPermission(getCurrentUser(), Resource.WHITE_LABELING, operation);
    }

}

File: ui/src/app/admin/custom-translation.controller.js
/*
 * ThingsBoard, Inc. ("COMPANY") CONFIDENTIAL
 *
 * Copyright © 2016-2019 ThingsBoard, Inc. All Rights Reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of ThingsBoard, Inc. and its suppliers,
 * if any.  The intellectual and technical concepts contained
 * herein are proprietary to ThingsBoard, Inc.
 * and its suppliers and may be covered by U.S. and Foreign Patents,
 * patents in process, and are protected by trade secret or copyright law.
 *
 * Dissemination of this information or reproduction of this material is strictly forbidden
 * unless prior written permission is obtained from COMPANY.
 *
 * Access to the source code contained herein is hereby forbidden to anyone except current COMPANY employees,
 * managers or contractors who have executed Confidentiality and Non-disclosure agreements
 * explicitly covering such access.
 *
 * The copyright notice above does not evidence any actual or intended publication
 * or disclosure  of  this source code, which includes
 * information that is confidential and/or proprietary, and is a trade secret, of  COMPANY.
 * ANY REPRODUCTION, MODIFICATION, DISTRIBUTION, PUBLIC  PERFORMANCE,
 * OR PUBLIC DISPLAY OF OR THROUGH USE  OF THIS  SOURCE CODE  WITHOUT
 * THE EXPRESS WRITTEN CONSENT OF COMPANY IS STRICTLY PROHIBITED,
 * AND IN VIOLATION OF APPLICABLE LAWS AND INTERNATIONAL TREATIES.
 * THE RECEIPT OR POSSESSION OF THIS SOURCE CODE AND/OR RELATED INFORMATION
 * DOES NOT CONVEY OR IMPLY ANY RIGHTS TO REPRODUCE, DISCLOSE OR DISTRIBUTE ITS CONTENTS,
 * OR TO MANUFACTURE, USE, OR SELL ANYTHING THAT IT  MAY DESCRIBE, IN WHOLE OR IN PART.
 */
/*@ngInject*/
export default function CustomTranslationController($scope, $translate, utils, toast, types, securityTypes, userPermissionsService, customTranslationService) {

    var vm = this;

    vm.readonly = !userPermissionsService.hasGenericPermission(securityTypes.resource.whiteLabeling, securityTypes.operation.write);

    vm.types = types;
    vm.customTranslation = {};
    vm.customTranslation.translationMap = {};
    vm.save = save;

    vm.downloadLocaleJson = downloadLocaleJson;

    vm.languageList = SUPPORTED_LANGS; //eslint-disable-line

    vm.currentLang = vm.languageList[0];

    vm.translationPlaceholder =
        '******* Example of custom translation ******** \n\n' +
        angular.toJson(
            {
                "home": {
                    "home": "My Home"
                },
                "custom": {
                    "my-dashboard": {
                        "title": "This is custom dashboard title",
                        "comment": "You can use it in your dashboard title using the following expression: {i18n:custom.my-dashboard.title}"
                    },
                    "my-widget": {
                        "legend-text": "This is custom legend text",
                        "comment": "You can use it in your dashboard widgets using the following expression: {i18n:custom.my-widget.legend-text}"
                    }
                }
            },
            2
        );

    vm.showError = function (error) {
        var toastParent = angular.element('#tb-custom-translation-panel');
        toast.showError(error, toastParent, 'top left');
    };

    getCurrentCustomTranslation();

    function getCurrentCustomTranslation() {
        var loadPromise = customTranslationService.getCurrentCustomTranslation();
        loadPromise.then(
            function success(customTranslation) {
                vm.customTranslation = customTranslation;
            });
    }

    function validate() {
        for (var lang in vm.customTranslation.translationMap) {
            var translation = vm.customTranslation.translationMap[lang];
            if (translation) {
                try {
                    angular.fromJson(translation);
                } catch (e) {
                    var details = utils.parseException(e);
                    var errorInfo = 'Error parsing JSON for ' + $translate.instant('language.locales.' + lang) + ' language:';
                    if (details.name) {
                        errorInfo += ' ' + details.name + ':';
                    }
                    if (details.message) {
                        errorInfo += ' ' + details.message;
                    }
                    vm.showError(errorInfo);
                    return false;
                }
            } else {
                delete vm.customTranslation.translationMap[lang];
            }
        }
        return true;
    }

    function save() {
        if (validate()) {
            var savePromise = customTranslationService.saveCustomTranslation(vm.customTranslation);
            savePromise.then(
                function success() {
                    vm.customTranslationForm.$setPristine();
                });
        }
    }

    function downloadLocaleJson() {
        customTranslationService.downloadLocaleJson(vm.currentLang);
    }
}